src/Symfony/Bridge/Doctrine/Validator/Constraints/UniqueEntityValidator.php

```diff
@@ -57,36 +58,60 @@ public function validate($entity, Constraint $constraint)
             throw new ConstraintDefinitionException('At least one field has to be specified.');
         }
 
-        if (null === $entity) {
+        if (null === $object) {
             return;
         }
 
+        $entityClass = \get_class($object);
+
         if ($constraint->em) {
             $em = $this->registry->getManager($constraint->em);
 
             if (!$em) {
                 throw new ConstraintDefinitionException(sprintf('Object manager "%s" does not exist.', $constraint->em));
             }
         } else {
-            $em = $this->registry->getManagerForClass(\get_class($entity));
+            if (null !== $constraint->entityClass) {
+                $em = $this->registry->getManagerForClass($constraint->entityClass);
+            } else {
+                $em = $this->registry->getManagerForClass($entityClass);
+            }
 
             if (!$em) {
-                throw new ConstraintDefinitionException(sprintf('Unable to find the object manager associated with an entity of class "%s".', get_debug_type($entity)));
+                throw new ConstraintDefinitionException(sprintf('Unable to find the object manager associated with an entity of class "%s".', get_debug_type($object)));
             }
         }
 
-        $class = $em->getClassMetadata(\get_class($entity));
+        try {
+            $repository = $em->getRepository($entityClass);
+            $isEntity = true;
+        }catch (\Exception $e) {
```

Comment:
Can we be more specific about the exception that we expect here?

Reply:
The problem is that in the test environment the returned error is:  
`Doctrine\ORM\Mapping\MappingException`  
but normally it is:  
`Doctrine\Persistence\Mapping\MappingException;`  
Their lowest common ancestor is `Exception`.  
Do you have some suggestions?

Reply:
You can catch a union of both exceptions.

```php
} catch (ExceptionOne | ExceptionTwo $e) {
```

Reply:
Yes. To be more precise, I didn't like the idea to write not needed code, so it would pass the tests.   But if this is acceptable, I will change it.  

P.S. fabbot.o don't understand PHP attributes  
P.S.2 Some, not related test is failing (randomly?)

Reply:
Fixed